app/code/Magento/InventoryCatalogSearch/Model/Search/FilterMapper/StockStatusFilter.php

```diff
@@ -0,0 +1,250 @@
+        $this->indexNameBuilder = $indexNameBuilder;
+        $this->indexNameResolver = $indexNameResolver;
+    }
+
+    /**
+     * Adds filter by stock status to base select
+     *
+     * @param Select $select
+     * @param mixed $stockValues
+     * @param string $type
+     * @param bool $showOutOfStockFlag
+     *
+     * @return Select
+     *
+     * @throws LocalizedException
+     * @throws \InvalidArgumentException
+     */
+    public function apply(Select $select, $stockValues, $type, $showOutOfStockFlag)
+    {
+        if (!$this->isValidType($type)) {
+            throw new \InvalidArgumentException(sprintf('Invalid filter type: %s', $type));
+        }
+
+        $select = clone $select;
+        $mainTableAlias = $this->extractTableAliasFromSelect($select);
+
+        if ($type === StockStatusFilterInterface::FILTER_JUST_ENTITY ||
+            $type === StockStatusFilterInterface::FILTER_ENTITY_AND_SUB_PRODUCTS
+        ) {
+            $this->addProductEntityJoin($select, $mainTableAlias);
+            $this->addInventoryStockJoin($select, $showOutOfStockFlag);
+        }
+
+        if ($type === StockStatusFilterInterface::FILTER_JUST_SUB_PRODUCTS ||
+            $type === StockStatusFilterInterface::FILTER_ENTITY_AND_SUB_PRODUCTS
+        ) {
+            $this->addSubProductEntityJoin($select, $mainTableAlias);
+            $this->addSubProductInventoryStockJoin($select, $showOutOfStockFlag);
+        }
+
+        return $select;
+    }
+
+    /**
+     * @param string $type
+     * @return bool
+     */
+    private function isValidType($type): bool
+    {
+        return \in_array($type, [
+            static::FILTER_JUST_SUB_PRODUCTS,
+            static::FILTER_JUST_ENTITY,
+            static::FILTER_ENTITY_AND_SUB_PRODUCTS
+        ], true);
+    }
+
+    /**
+     * @param Select $select
+     * @param string $mainTableAlias
+     */
+    private function addProductEntityJoin(Select $select, $mainTableAlias)
+    {
+        $select->joinInner(
+            [static::TABLE_ALIAS_PRODUCT => 'catalog_product_entity'],
+            sprintf('%s.entity_id = %s.entity_id', static::TABLE_ALIAS_PRODUCT, $mainTableAlias),
+            []
+        );
+    }
+
+    /**
+     * @param Select $select
+     * @param string $mainTableAlias
+     */
+    private function addSubProductEntityJoin(Select $select, $mainTableAlias)
+    {
+        $select->joinInner(
+            [static::TABLE_ALIAS_SUB_PRODUCT => 'catalog_product_entity'],
+            sprintf('%s.entity_id = %s.source_id', static::TABLE_ALIAS_SUB_PRODUCT, $mainTableAlias),
+            []
+        );
+    }
+
+    /**
+     * @param Select $select
+     * @param bool $showOutOfStockFlag
+     * @throws LocalizedException
+     */
+    private function addInventoryStockJoin(Select $select, $showOutOfStockFlag)
+    {
+        $select->joinInner(
+            [static::TABLE_ALIAS_STOCK_INDEX => $this->getStockTableName()],
```

Comment:
why do you use `static`  for access to all constants? 
are you expecting to use late static binding here? 
but constants could not be overloaded